plugins/inputs/gnmi/gnmi.go

```diff
@@ -292,22 +315,65 @@ func (c *GNMI) handleSubscribeResponseUpdate(address string, response *gnmiLib.S
 
 	if response.Update.Prefix != nil {
 		var err error
-		if prefix, prefixAliasPath, err = c.handlePath(response.Update.Prefix, prefixTags, ""); err != nil {
+		if prefix, prefixAliasPath, err = c.handlePath(response.Update.Prefix, prefixTags, "", true); err != nil {
 			c.Log.Errorf("handling path %q failed: %v", response.Update.Prefix, err)
 		}
 	}
 	prefixTags["source"], _, _ = net.SplitHostPort(address)
 	prefixTags["path"] = prefix
 
+	// Prepare tags from prefix
+	tags := make(map[string]string, len(prefixTags))
+	for key, val := range prefixTags {
+		tags[key] = val
+	}
+
 	// Parse individual Update message and create measurements
 	var name, lastAliasPath string
-	for _, update := range response.Update.Update {
-		// Prepare tags from prefix
-		tags := make(map[string]string, len(prefixTags))
-		for key, val := range prefixTags {
-			tags[key] = val
+	var aliasPath string
+
+	tagsPerUpdate := make(map[int]map[string]string)
+	fieldsPerUpdate := make(map[int]map[string]interface{})
+
+	// Extract the tags and embedded tags first
+	assumeExtraTags := false
+	for i, update := range response.Update.Update {
+		tagsPerUpdate[i] = make(map[string]string)
+
+		var hasExtraTags bool
+		_, _, hasExtraTags = c.handleTelemetryField(update, tagsPerUpdate[i], prefix, true)
+		if hasExtraTags {
+			assumeExtraTags = true
+		}
+	}
+
+	// If there were no tags specified with the response,
+	// copy the embedded tags, if any, into tags
+	if assumeExtraTags {
+		for _, t := range tagsPerUpdate {
+			for k, v := range t {
+				tags[k] = v
+			}
+		}
+	} else {
+		// Copy the global tags into the tags for this update
+		for _, t := range tagsPerUpdate {
+			for k, v := range tags {
+				t[k] = v
+			}
+		}
+	}
+
+	// Extract the fields and add them to the grouper together with the tags
+	for i, update := range response.Update.Update {
+		fieldsPerUpdate[i] = make(map[string]interface{})
```

Comment:
Is `fieldsPerUpdate` ever used outside of this loop? If not, you can just use a _single_ local variable instead of carrying the array.